src/Cardano/AddressDerivation/Byron.hs

```diff
@@ -0,0 +1,252 @@
+
+blake2b256 :: ScrubbedBytes -> ScrubbedBytes
+blake2b256 = BA.convert . hash @ScrubbedBytes @Blake2b_256
+
+-- | Derive a symmetric key for encrypting and authenticating the address
+-- derivation path. PBKDF2 encryption using HMAC with the hash algorithm SHA512
+-- is employed.
+hdPassphrase :: XPub -> ScrubbedBytes
+hdPassphrase masterKey =
+    PBKDF2.generate
+    (PBKDF2.prfHMAC SHA512)
+    (PBKDF2.Parameters 500 32)
+    (unXPub masterKey)
+    ("address-hashing" :: ByteString)
+
+mkByronKeyFromMasterKey
+    :: XPrv
+    -> ByronKey 'RootK XPrv
+mkByronKeyFromMasterKey = unsafeMkByronKeyFromMasterKey ()
+
+unsafeMkByronKeyFromMasterKey
+    :: DerivationPath depth
+    -> XPrv
+    -> ByronKey depth XPrv
+unsafeMkByronKeyFromMasterKey derivationPath masterKey = ByronKey
+    { getKey = masterKey
+    , derivationPath
+    , payloadPassphrase = hdPassphrase (toXPub masterKey)
+    }
+
+{-------------------------------------------------------------------------------
+                                 HD derivation
+-------------------------------------------------------------------------------}
+
+-- | Derives account private key from the given root private key, using
+-- derivation scheme 1.
+deriveAccountPrivateKey
+    :: ScrubbedBytes
+    -> ByronKey 'RootK XPrv
+    -> Index 'WholeDomain 'AccountK
+    -> ByronKey 'AccountK XPrv
+deriveAccountPrivateKey pwd masterKey idx@(Index accIx) = ByronKey
+    { getKey = deriveXPrv DerivationScheme1 pwd (getKey masterKey) accIx
+    , derivationPath = idx
+    , payloadPassphrase = payloadPassphrase masterKey
+    }
+
+-- | Derives address private key from the given account private key, using
+-- derivation scheme 1.
+deriveAddressPrivateKey
+    :: ScrubbedBytes
+    -> ByronKey 'AccountK XPrv
+    -> Index 'WholeDomain 'AddressK
+    -> ByronKey 'AddressK XPrv
+deriveAddressPrivateKey pwd accountKey idx@(Index addrIx) = ByronKey
+    { getKey = deriveXPrv DerivationScheme1 pwd (getKey accountKey) addrIx
+    , derivationPath = (derivationPath accountKey, idx)
+    , payloadPassphrase = payloadPassphrase accountKey
+    }
+
+{-------------------------------------------------------------------------------
+                                     Utils
+-------------------------------------------------------------------------------}
+
+-- | Checks whether or not an invariant holds, by applying the given predicate
+--   to the given value.
+--
+-- If the invariant does not hold (indicated by the predicate function
+-- returning 'False'), throws an error with the specified message.
+--
+-- >>> invariant "not empty" [1,2,3] (not . null)
+-- [1, 2, 3]
+--
+-- >>> invariant "not empty" [] (not . null)
+-- *** Exception: not empty
+invariant
+    :: String
+        -- ^ The message
+    -> a
+        -- ^ The value to test
+    -> (a -> Bool)
+        -- ^ The predicate
+    -> a
+invariant msg a predicate =
+    if predicate a then a else error msg
```

Comment:
What about using: https://hackage.haskell.org/package/base-4.6.0.1/docs/Control-Exception-Base.html#v:assert instead? It's built-in, and raises proper exceptions.

Reply:
corrected in https://github.com/input-output-hk/cardano-addresses/pull/4/commits/06ab4de0503ee77462630ceb6a7b4f195fb80a75